### PR TITLE
Issue #12553: Add 'when' to default IllegalIdentifierName pattern

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #12553: Add 'when' to default IllegalIdentifierName 

------------------------------------------------------

Diff Regression config: https://gist.githubusercontent.com/Kevin222004/f8cbb850ce72234e08c6963c6f4147b6/raw/25d74a45948b256b83ba02fb72a9882214916120/name.xml

### CODE REVIEW

**Comment by Kevin222004** [comment]
I have remove this  test case from previous latest commit to avoid ci failure because this is not compilable with java19 i will add after resolving this
```
    void m1(Object o) {
        switch (o) {
            case Integer when when i1 == 0: // violation
            case Integer noWhen when i1 == 1: // ok
                m2(i1);
                break;
        }
    }
    
    void m2() {}
```

**Comment by Kevin222004** [comment]
GitHub, generate report

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameWhen.java
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameWhen.java:23]: add message to all "// violation"

Do you any suggestion on how to enforce this rule by Checkstyle and put in suppression all existing files, to let us violate new files.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Comment by Kevin222004** [comment]
GitHub, generate report

**Comment by Kevin222004** [comment]
GitHub, generate report

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameWhen.java
Comment (Kevin222004) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameWhen.java:23]: it is becoming difficult to put violation message here bcs msg is too long if we use violation above and below still message become longer than 100 line, 
although if we not consider the 100 line it is failing only with 
```
Actual and expected violations differ.
expected to match: 16:(?:\d+:)?\s.*Name 'When' must match pattern '\(?i\)^\(?!\(record|yield|var|permits|sealed|when|_\)$\).+$'.*
but was          : 16:9: Name 'When' must match pattern '(?i)^(?!(record|yield|var|permits|sealed|when|_)$).+$'.
```  

we can write the message here something like this 
`    int When = 10; // violation 'Name 'When' must match pattern .*' ` this is working fine is it good to show message like this 

> Do you any suggestion on how to enforce this rule by Checkstyle and put in suppression all existing files, to let us violate new files.

i have to think on this right know i have not any suggestion once the issue https://github.com/checkstyle/checkstyle/issues/11214 will resolve and all the suppression from bdd is removed then if we not write the violation message on that file the mvn it would not pass 
right know  i want have any idea

Comment (Kevin222004) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/naming/illegalidentifiername/InputIllegalIdentifierNameWhen.java]: AS of know message written

**Comment by romani** [comment]
Code is good, but report is full of noise only due to change in message.
Please update config to override message to make it as it was before without 'when'.
https://checkstyle.org/config.html#Custom_messages

**Comment by Kevin222004** [comment]
GitHub, generate report

**Comment by Kevin222004** [comment]
> Code is good, but report is full of noise only due to change in message. Please update config to override message to make it as it was before without 'when'. https://checkstyle.org/config.html#Custom_messages

done

**Review by romani — APPROVED** [review]
ok to merge .

But I have bad feeling that a lot of items become to be violation. This might be a sign that it is not good to make it behavior by default.
Lets ask other maintainers for opinion.

**Comment by nrmancuso** [comment]
> But I have bad feeling that a lot of items become to be violation. This might be a sign that it is not good to make it behavior by default.

After looking at the reports and thinking on this, I tend to agree. Maybe we should wait on this update at least until we are sure that `when` will make it into Java officially. I was not aware that `when` was such a common naming convention. Also, considering that `when` is limited to only case labels, it may not be such a bad practice to use it elsewhere.

**Comment by Kevin222004** [comment]
can't we make it as a separate property know

**Comment by rnveach** [comment]
User can always turn off what conflicts for them.

https://checkstyle.org/config_naming.html#IllegalIdentifierName
> Checks identifiers with a pattern for a set of illegal names, such as those that are restricted or contextual keywords. 

I can't believe we never ran into this for `record`.

> wait on this update at least until we are sure that when will make it into Java officially

I atleast agree with this. This is not a change in our grammar, so we don't need to worry about failed parsing. My mindset is unofficial Java things are basically just test grounds and they could come and go. It seems odd to support something like this officially from us when it is not even official for Java.

I am more for not supporting unofficial things, but if we want, we can atleast add documentation for users on how to add it if they wish.

**Comment by Kevin222004** [comment]
What would be the next step, should we wait for the official update of java for when

**Comment by rnveach** [comment]
@nrmancuso @romani ping. We need clarification on issue direction.

**Comment by romani** [comment]
We can not merge it until `when` decome a final and official. For now it is in-preview feature of jdk, so might change.

**Comment by romani** [comment]
So others also does not like this update as it affects existing code bases too much.
So having it by default will be too aggressive.

After thinking on this, we should not do this as default behavior. But if no reason to work by default, no reason in this update.

@Kevin222004 , sorry that we wasted your time

**Pull request closed without merging**
